homeassistant/components/goodwe/sensor.py

```diff
@@ -0,0 +1,252 @@
+    )
+
+    return True
+
+
+class InverterEntity(CoordinatorEntity, SensorEntity):
+    """Entity representing the inverter instance itself."""
+
+    _MAIN_ENTITY_SENSOR = "ppv"
+
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator,
+        device_info: DeviceInfo,
+        inverter: Inverter,
+    ) -> None:
+        """Initialize the main inverter entity."""
+        super().__init__(coordinator)
+        self.entity_id = f".{DOMAIN}_inverter"
+        self._attr_unique_id = f"{DOMAIN}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+        self._attr_icon = "mdi:solar-power"
+        self._attr_name = "PV Inverter"
+        self._attr_native_unit_of_measurement = POWER_WATT
+        self._attr_extra_state_attributes = {"serial_number": inverter.serial_number}
+        self._inverter: Inverter = inverter
+        self._previous_value = None
+
+    async def async_set_work_mode(self, work_mode: int) -> None:
+        """Set the inverter work mode."""
+        await self._inverter.set_work_mode(work_mode)
+
+    async def async_set_ongrid_battery_dod(self, ongrid_battery_dod: int) -> None:
+        """Set the on-grid battery dod."""
+        await self._inverter.set_ongrid_battery_dod(ongrid_battery_dod)
+
+    async def async_set_grid_export_limit(self, grid_export_limit: int) -> None:
+        """Set the grid export limit."""
+        await self._inverter.set_grid_export_limit(grid_export_limit)
+
+    @property
+    def native_value(self):
+        """Return the value reported by the sensor."""
+        new_value = self.coordinator.data.get(
+            self._MAIN_ENTITY_SENSOR, self._previous_value
+        )
+        self._previous_value = new_value
+        return new_value
+
+    @property
+    def state_attributes(self):
+        """Return the inverter state attributes."""
+        return self.coordinator.data
+
+
+class InverterSensor(CoordinatorEntity, SensorEntity):
+    """Entity representing individual inverter sensor."""
+
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator,
+        device_info: DeviceInfo,
+        inverter: Inverter,
+        sensor: Sensor,
+    ) -> None:
+        """Initialize an inverter sensor."""
+        super().__init__(coordinator)
+        self.entity_id = f".{DOMAIN}_{sensor.id_}"
+        self._attr_unique_id = f"{DOMAIN}-{sensor.id_}-{inverter.serial_number}"
+        self._attr_name = sensor.name.strip()
+        self._attr_device_info = device_info
+        self._attr_entity_category = (
+            ENTITY_CATEGORY_DIAGNOSTIC if sensor.id_ not in _MAIN_SENSORS else None
+        )
+        if sensor.unit in _DESCRIPTIONS:
+            self.entity_description = _DESCRIPTIONS[sensor.unit]
+        else:
+            self._attr_native_unit_of_measurement = sensor.unit
+        self._attr_icon = _ICONS.get(sensor.kind)
+        # Set the inverter SoC as main device battery sensor
+        if sensor.id_ == BATTERY_SOC:
```

Comment:
Why not move all these cases to different entity descriptions instead and have one description per sensor kind? That's much more readable.

Reply:
I'm not sure I understand your suggestion. There are already 7 kinds of entity descriptions.
Are you referring to the `self._attr_icon = _ICONS.get(sensor.kind)` ?
That's intentionally not in the description since there are often 3 variants of these entity types (AC/UPS/BAT) and that would lead to having 21 different entity descriptions, which is probably not helping readability.

Reply:
It's much more readable to have many descriptions, by using a declarative approach, than using conditional logic here.

I'd create one description per measurment kind.

Reply:
I see, but I'm not sure fully declarative approach is feasible.
I can remove the `self._attr_icon = _ICONS.get(sensor.kind)` single line.
(at cost of creating ~12 new "declarative" entries for the _DESCRIPTIONS and using (unit, kind) tuple keys there).
But the conditional lookup via `if sensor.unit in _DESCRIPTIONS` can't really be avoided.

The trouble is that there is not fixed set of sensors to be listed declaratively. 
The set of sensors heavily depends on family, model and enabled capabilities of individual inverters. 
And the unified set of all potential sensors is a large one, more than 150 sensors are out there.
I'm not sure having them copy/pasted from the inverter communication library just to have them statically declared would help readability. 
In any way, dynamic lookup through supported sensors would be needed and so some form of lookup/filter like the
```
if sensor.unit in _DESCRIPTIONS:
        self.entity_description = _DESCRIPTIONS[sensor.unit]
```
would need to be there anyway.

Unless I missed or misunderstood something, the only potential improvement here is removal of single line at cost of creating 60 copy/pasted ones.

Reply:
Are there 150 different variations of sensors or are there overlap so that more than one sensor would have the same description?

Reply:
No, there are 7 major types of sensors (which are declared already in _DESCRIPTIONS dict). 
And for many of those sensors, the are 3 variants - AC/UPS/BAT - e.g.  voltage L1 AC, voltage L1 UPS, voltage BAT.
Each sensor knows its unit (7 options) and kind (3 options) (or has none of those).
So yes, we can have instead of existing 7 declared entity descriptions some 15-20 (need to check the underlying library, the temperature most like have only 2 kinds).
But that would remove just that single line settings the icon from the sensor kind.
I'm always willing to make the code more readable, but in this case I don't see how, or I did not understand what you suggest.

Reply:
I was thinking we can have a dict that maps the sensor ids to the description. The keys in the dict can be tuples of sensor ids, so that more than one sensor id maps to the same description. We can iterate over the dict and look up the correct description by checking if the sensor id is in the key tuple. This allows for both the main 15-20 kinds and some specific cases like `BATTERY_SOC`.

Reply:
Sensor IDs are not know, resp. there are lots of them and they are dynamic. 
They are characterized by (unit, kind) and each such tuple can share the same description.
At the moment there 7 descriptions in the dict, keyed by unit and the kind is used on top of that for the icon.
At his moment we are iterating over the dict (not by sensor id but by sensor unit).
We can iterate and lookup by (unit, kind) instead (and create descriptions W_AC, W_UPS, W_PV, W_BAT, W_GRID ...).
I was wrong above, there are not 3 but 5 kinds of sensors, so we may have like 25-30 of those descriptions.
But all we achieve is removal of line 233, nothing more.

Reply:
Ok, if we don't know the different ids we can leave it as it is for now.